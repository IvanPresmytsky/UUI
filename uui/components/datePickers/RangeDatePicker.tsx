import React, { useState } from 'react';
import cx from 'classnames';
import { uuiMod, DropdownBodyProps, devLogger, withMods, IDropdownTogglerProps } from '@epam/uui-core';
import { Dropdown } from '@epam/uui-components';
import { DropdownContainer } from '../overlays';
import { FlexRow } from '../layout';
import { RangeDatePickerBody } from './RangeDatePickerBody';
import { TextInput } from '../inputs';
import { systemIcons } from '../../icons/icons';
import { i18n } from '../../i18n';
import css from './RangeDatePicker.module.scss';
import { defaultValue, useRangeDatePickerState } from './useRangeDatePickerState';
import { RangeDatePickerProps } from './types';
import dayjs from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat.js';

dayjs.extend(customParseFormat);

const modifiers = [{
    name: 'offset',
    options: { offset: [0, 6] },
}];

function RangeDatePickerComponent(props: RangeDatePickerProps): JSX.Element {
    const [isOpen, setIsOpen] = useState(false);

    const {
        state,
        onRangeChange,
        clearRange,
        handleBlur,
        handleFocus,
        toggleIsOpen,
        getChangeHandler,
    } = useRangeDatePickerState({
        ...props,
        onOpenChange: (value: boolean) => {
            setIsOpen(value);
            props.onOpenChange?.(value);
        },
    });

    const renderBody = (renderProps: DropdownBodyProps): JSX.Element => {
        if (!props.isReadonly && !props.isDisabled) {
            const value = {
                selectedDate: props.value || defaultValue,
                month: state.month,
                view: state.view,
                activePart: state.inFocus,
            };

            return (
                <DropdownContainer
                    { ...renderProps }
                    cx={ cx(css.dropdownContainer) }
                    focusLock={ false }
                >
                    <FlexRow>
                        <RangeDatePickerBody
                            cx={ cx(props.bodyCx) }
                            value={ value }
                            onValueChange={ onRangeChange }
                            filter={ props.filter }
                            presets={ props.presets }
                            renderDay={ props.renderDay }
                            renderFooter={ () => {
                                return props.renderFooter?.(props.value || defaultValue);
                            } }
                            isHoliday={ props.isHoliday }
                            rawProps={ props.rawProps?.body }
                        />
                    </FlexRow>
                </DropdownContainer>
            );
        }
    };

    const renderInput = (renderProps: IDropdownTogglerProps): JSX.Element => {
        if (__DEV__) {
            if (props.size === '48') {
                devLogger.warnAboutDeprecatedPropValue<RangeDatePickerProps, 'size'>({
                    component: 'RangeDatePicker',
                    propName: 'size',
                    propValue: props.size,
                    propValueUseInstead: '42',
                    condition: () => ['48'].indexOf(props.size) !== -1,
                });
            }
        }

        const clearAllowed = !props.disableClear && state.inputValue.from && state.inputValue.to;

        return (
            <div
                className={ cx(
                    props.inputCx,
                    css.dateInputGroup,
                    props.isDisabled && uuiMod.disabled,
                    props.isReadonly && uuiMod.readonly,
                    props.isInvalid && uuiMod.invalid,
                    state.inFocus && uuiMod.focus,
                ) }
                onClick={ !props.isDisabled && renderProps.onClick }
                // onBlur={ handleWrapperBlur }
                ref={ renderProps.ref }
            >
                <TextInput
                    icon={ systemIcons[props.size || '36'].calendar }
                    cx={ cx(css.dateInput, css['size-' + (props.size || 36)], state.inFocus === 'from' && uuiMod.focus) }
                    size={ props.size || '36' }
                    placeholder={ props.getPlaceholder ? props.getPlaceholder('from') : i18n.rangeDatePicker.pickerPlaceholderFrom }
                    value={ state.inputValue.from }
                    onValueChange={ getChangeHandler('from') }
                    onFocus={ (event) => handleFocus(event, 'from') }
                    onBlur={ (event) => handleBlur(event, 'from') }
                    ///
                    isInvalid={ props.isInvalid }
                    isDisabled={ props.isDisabled }
                    isReadonly={ props.isReadonly }
                    isDropdown={ false }
                    rawProps={ props.rawProps?.from }
                    id={ props?.id }
                />
                <div className={ css.separator } />
                <TextInput
                    cx={ cx(css.dateInput, css['size-' + (props.size || 36)], state.inFocus === 'to' && uuiMod.focus) }
                    placeholder={ props.getPlaceholder ? props.getPlaceholder('to') : i18n.rangeDatePicker.pickerPlaceholderTo }
                    size={ props.size || '36' }
                    value={ state.inputValue.to }
                    onCancel={ clearAllowed && clearRange }
                    onValueChange={ getChangeHandler('to') }
                    onFocus={ (e) => handleFocus(e, 'to') }
                    onBlur={ (e) => handleBlur(e, 'to') }
                    ///
                    isInvalid={ props.isInvalid }
                    isDisabled={ props.isDisabled }
                    isReadonly={ props.isReadonly }
                    isDropdown={ false }
                    rawProps={ props.rawProps?.to }
                />
            </div>
        );
    };

    return (
        <Dropdown
            renderTarget={ (renderProps) => {
                return props.renderTarget?.(renderProps) || renderInput(renderProps);
            } }
            renderBody={ (renderProps) => renderBody(renderProps) }
            onValueChange={ toggleIsOpen }
            value={ isOpen }
            modifiers={ modifiers }
            placement={ props.placement }
            forwardedRef={ props.forwardedRef }
        />
    );
}

export const RangeDatePicker = withMods(RangeDatePickerComponent);
