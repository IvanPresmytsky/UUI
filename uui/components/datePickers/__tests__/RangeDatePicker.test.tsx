import * as React from 'react';
import { RangeDatePicker, RangeDatePickerProps } from '../RangeDatePicker';
import {
    renderSnapshotWithContextAsync, setupComponentForTest, fireEvent, screen, within, userEvent,
} from '@epam/uui-test-utils';

interface TestProps {
    value?: RangeDatePickerProps['value'];
    format?: RangeDatePickerProps['format'];
    onBlur?: RangeDatePickerProps['onBlur'];
    onFocus?: RangeDatePickerProps['onFocus'];
    isHoliday?: RangeDatePickerProps['isHoliday'];
    onOpenChange?: RangeDatePickerProps['onOpenChange'];
}

function parentElemContainsClasses(elem: HTMLElement, classesArr: string[]) {
    // @ts-ignore
    const actualList = [...elem.parentElement.classList];
    return classesArr.every((c: string) => actualList.indexOf(c) !== -1);
}

async function setupRangeDatePicker(props: TestProps) {
    const { result, mocks } = await setupComponentForTest<RangeDatePickerProps>(
        (context) => ({
            rawProps: { from: { 'data-testid': 'from' }, to: { 'data-testid': 'to' } },
            ...props,
            value: props.value || { from: null, to: null },
            onValueChange: jest.fn().mockImplementation((newValue) => {
                context.current?.setProperty('value', newValue);
            }),
            size: '48',
        }),
        (props) => <RangeDatePicker { ...props } />,
    );

    const from = within(screen.getByTestId('from')).getByRole<HTMLInputElement>('textbox');
    const to = within(screen.getByTestId('to')).getByRole<HTMLInputElement>('textbox');

    return {
        result,
        dom: { from, to },
        mocks: { onValueChange: mocks.onValueChange, onOpenChange: mocks.onOpenChange },
    };
}

describe('RangeDataPicker', () => {
    it('should be rendered if minimum params and custom format defined', async () => {
        const tree = await renderSnapshotWithContextAsync(<RangeDatePicker format="MMM D, YYYY" value={ { from: null, to: null } } onValueChange={ jest.fn } />);
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered if many params defined', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <RangeDatePicker
                format="MMM D, YYYY"
                value={ { from: null, to: null } }
                onValueChange={ jest.fn }
                renderFooter={ ((value: any) => jest.fn(value)) as any }
                disableClear={ false }
                getPlaceholder={ () => '' }
                isDisabled
                isReadonly
                isInvalid
            />,
        );
        expect(tree).toMatchSnapshot();
    });

    it('should change input value after change props', async () => {
        const value = { from: '2017-01-22', to: '2017-01-28' };
        const { dom } = await setupRangeDatePicker({ value });
        expect(dom.from.value).toBe('Jan 22, 2017');
        expect(dom.to.value).toBe('Jan 28, 2017');
    });

    it('should render with default props', async () => {
        const { result } = await setupRangeDatePicker({ value: undefined });
        expect(result.container).not.toBeFalsy();
    });

    it('should change state on picker clear', async () => {
        const value = { from: '2017-01-22', to: '2017-01-28' };
        const { mocks } = await setupRangeDatePicker({ value });
        const clear = screen.getByRole('button');
        fireEvent.click(clear);
        expect(mocks.onValueChange).toBeCalledWith({ from: null, to: null });
    });

    it("should open picker on 'from' field focus and close it on blur", async () => {
        const { dom } = await setupRangeDatePicker({ value: undefined });
        fireEvent.focus(dom.from);
        expect(screen.getByRole('dialog')).toBeInTheDocument();
        fireEvent.blur(dom.from);
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('should open picker on "To" field focus and close it on blur', async () => {
        const { dom } = await setupRangeDatePicker({ value: undefined });
        fireEvent.focus(dom.to);
        expect(screen.getByRole('dialog')).toBeInTheDocument();
        fireEvent.blur(dom.to);
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('should reset invalid "From" value onBlur', async () => {
        const value = { from: '2019-10-47', to: '2019-10-07' };
        const { dom, mocks } = await setupRangeDatePicker({ value });
        fireEvent.blur(dom.from);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith({
            from: null,
            to: value.to,
        });
    });

    it('should set new value when new value typed in input', async () => {
        const newValueManualTyping = { from: 'Sep 11, 2019', to: 'Sep 20, 2019' };
        const value = { from: '2019-09-14', to: '2019-09-15' };
        const { dom, mocks } = await setupRangeDatePicker({ value });

        fireEvent.change(dom.from, { target: { value: newValueManualTyping.from } });
        fireEvent.blur(dom.from);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith({
            from: '2019-09-11',
            to: '2019-09-15',
        });
    });

    it('should change range on picker body value change', async () => {
        const newValueManualSel = { from: '2019-09-11' };
        const value = { from: '2019-09-10', to: '2019-09-12' };
        const { dom, mocks } = await setupRangeDatePicker({ value });

        fireEvent.focus(dom.from);
        const dialog = screen.getByRole('dialog');
        const [sept11] = await within(dialog).findAllByText('11');
        fireEvent.click(sept11);
        expect(mocks.onValueChange).toHaveBeenLastCalledWith({
            from: newValueManualSel.from,
            to: '2019-09-12',
        });
    });

    it('should focus corresponding input on picker values change', async () => {
        const value = { from: '2019-09-10', to: '2019-09-12' };
        const { dom } = await setupRangeDatePicker({ value });

        await userEvent.click(dom.from);
        expect(parentElemContainsClasses(dom.from, ['uui-focus'])).toBeTruthy();
        expect(parentElemContainsClasses(dom.to, ['uui-focus'])).toBeFalsy();

        const dialog = screen.getByRole('dialog');
        const [, oct11] = await within(dialog).findAllByText('11');

        expect(dom.from.value).toBe('Sep 10, 2019');
        expect(dom.to.value).toBe('Sep 12, 2019');

        // on select, focus should be moved to 'to' input
        await userEvent.click(oct11);
        expect(parentElemContainsClasses(dom.from, ['uui-focus'])).toBeFalsy();
        expect(parentElemContainsClasses(dom.to, ['uui-focus'])).toBeTruthy();
        expect(dom.from.value).toBe('Oct 11, 2019');
        expect(dom.to.value).toBe('');

        const [, oct5] = await within(dialog).findAllByText('5');

        // since earlier date is selected, focus should be on 'to' elem still
        await userEvent.click(oct5);
        expect(parentElemContainsClasses(dom.from, ['uui-focus'])).toBeFalsy();
        expect(parentElemContainsClasses(dom.to, ['uui-focus'])).toBeTruthy();
        expect(dom.from.value).toBe('Oct 5, 2019');
        expect(dom.to.value).toBe('');

        const [, oct25] = await within(dialog).findAllByText('25');

        // should cancel focus when two dates selected
        await userEvent.click(oct25);
        expect(parentElemContainsClasses(dom.from, ['uui-focus'])).toBeFalsy();
        expect(parentElemContainsClasses(dom.to, ['uui-focus'])).toBeFalsy();
        expect(dom.from.value).toBe('Oct 5, 2019');
        expect(dom.to.value).toBe('Oct 25, 2019');
    });

    it('should format value onBlur', async () => {
        const value = { from: '2019-09-10', to: '2019-10-10' };
        const newValueInIsoFormat = { from: '2019-09-11', to: '2019-10-11' };

        const { dom } = await setupRangeDatePicker({ value });
        expect(dom.from.value).toBe('Sep 10, 2019');
        fireEvent.change(dom.from, { target: { value: newValueInIsoFormat.from } });
        fireEvent.blur(dom.from);
        expect(dom.from.value).toBe('Sep 11, 2019');

        fireEvent.change(dom.to, { target: { value: newValueInIsoFormat.to } });
        fireEvent.blur(dom.to);
        expect(dom.to.value).toBe('Oct 11, 2019');
    });

    it('should return format', async () => {
        const value = { from: '2019-09-10', to: '2019-10-10' };
        const { dom } = await setupRangeDatePicker({ value, format: 'DD-MM-YYYY' });
        expect(dom.from.value).toBe('10-09-2019');
        expect(dom.to.value).toBe('10-10-2019');
    });

    it('should set the same value on from: & to: input', async () => {
        const value = { from: '2019-09-10', to: '2019-09-10' };
        const { dom } = await setupRangeDatePicker({ value });
        expect(dom.from.value).toBe('Sep 10, 2019');
        expect(dom.to.value).toBe('Sep 10, 2019');
    });

    it('should fire onOpenChange event on open state change', async () => {
        const value = { from: '2017-01-22', to: '2017-01-28' };
        const onOpenChange = jest.fn();
        const { dom } = await setupRangeDatePicker({ value, onOpenChange });
        fireEvent.focus(dom.from);
        expect(onOpenChange).toBeCalledWith(true);
        fireEvent.blur(dom.from);
        expect(onOpenChange).toBeCalledWith(false);
    });

    it('should support entering from keyboard', async () => {
        const value = { from: '2019-09-10', to: '2019-09-15' };
        const onOpenChange = jest.fn();
        const onBlur = jest.fn();
        const onFocus = jest.fn();
        const { dom, result } = await setupRangeDatePicker({ value, onBlur, onFocus, onOpenChange });

        await userEvent.clear(dom.from);
        expect(onFocus).toBeCalled();
        expect(dom.from).toHaveFocus();
        expect(onOpenChange).toHaveBeenCalledWith(true);

        await userEvent.type(dom.from, '2019-09-11');

        await userEvent.click(result.container);
        expect(onBlur).toBeCalled();
        expect(dom.from).not.toHaveFocus();
        expect(onOpenChange).toHaveBeenCalledWith(false);

        expect(dom.from.value).toBe('Sep 11, 2019');
        expect(dom.to.value).toBe('Sep 15, 2019');
    });

    it('should support entering from keyboard with custom format', async () => {
        const value = { from: '2019-09-10', to: '2019-09-15' };
        const onOpenChange = jest.fn();
        const onBlur = jest.fn();
        const onFocus = jest.fn();
        const { result, dom } = await setupRangeDatePicker({
            value,
            format: 'DD-MM-YYYY',
            onBlur,
            onFocus,
            onOpenChange,
        });

        await userEvent.clear(dom.from);
        expect(onFocus).toBeCalled();
        expect(onOpenChange).toHaveBeenCalledWith(true);

        await userEvent.type(dom.from, '2019-09-11');
        expect(dom.from).toHaveFocus();

        await userEvent.click(result.container);
        expect(onBlur).toBeCalled();
        expect(dom.from).not.toHaveFocus();
        expect(onOpenChange).toHaveBeenCalledWith(false);

        expect(dom.from.value).toBe('11-09-2019');
        expect(dom.to.value).toBe('15-09-2019');
    });
});
