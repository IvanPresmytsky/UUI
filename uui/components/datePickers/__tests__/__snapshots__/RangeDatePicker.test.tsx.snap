// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeDataPicker should be rendered if many params defined 1`] = `
<div
  className="dateInputGroup uui-disabled uui-readonly uui-invalid"
  onBlur={[Function]}
  onClick={false}
>
  <div
    className="container uui-input-box uui-disabled uui-readonly uui-invalid root size-36 mode-form dateInput size-36"
    onFocus={[Function]}
    tabIndex={-1}
  >
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className=""
      />
    </div>
    <input
      aria-disabled={true}
      aria-invalid={true}
      className="uui-input"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      readOnly={true}
      tabIndex={-1}
      type="text"
      value=""
    />
  </div>
  <div
    className="separator"
  />
  <div
    className="container uui-input-box uui-disabled uui-readonly uui-invalid root size-36 mode-form dateInput size-36"
    onFocus={[Function]}
    tabIndex={-1}
  >
    <input
      aria-disabled={true}
      aria-invalid={true}
      className="uui-input"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      readOnly={true}
      tabIndex={-1}
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`RangeDataPicker should be rendered if minimum params and custom format defined 1`] = `
<div
  className="dateInputGroup"
  onBlur={[Function]}
  onClick={[Function]}
>
  <div
    className="container uui-input-box -clickable root size-36 mode-form dateInput size-36"
    onFocus={[Function]}
    tabIndex={-1}
  >
    <div
      className="container uui-icon uui-enabled"
      style={Object {}}
    >
      <svg
        className=""
      />
    </div>
    <input
      className="uui-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="From:"
      tabIndex={0}
      type="text"
      value=""
    />
  </div>
  <div
    className="separator"
  />
  <div
    className="container uui-input-box -clickable root size-36 mode-form dateInput size-36"
    onFocus={[Function]}
    tabIndex={-1}
  >
    <input
      className="uui-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="To:"
      tabIndex={0}
      type="text"
      value=""
    />
  </div>
</div>
`;
